src/reduced_order/pod_basis_offline.cpp

```diff
@@ -0,0 +1,156 @@
+                std::istringstream stream(line);
+                std::string field;
+                int col = 0;
+                while (getline(stream, field,' ')) { //parse data values on each line
+                    if (field.empty()) {
+                        continue;
+                    } else {
+                        snapshotMatrix(row, snapshotMatrix.cols()-cols+col) = std::stod(field); //This will work for however many solutions in each file
+                        col++;
+                    }
+                }
+                row++;
+            }
+            myfile.close();
+        }
+    }
+
+    pcout << "Snapshot matrix generated." << std::endl;
+
+
+    /* Reference for simple POD basis computation: Refer to Algorithm 1 in the following reference:
+    "Efficient non-linear model reduction via a least-squares Petrov–Galerkin projection and compressive tensor approximations"
+    Kevin Carlberg, Charbel Bou-Mosleh, Charbel Farhat
+    International Journal for Numerical Methods in Engineering, 2011
+    */
+
+    pcout << "Computing POD basis..." << std::endl;
+
+    VectorXd reference_state = snapshotMatrix.rowwise().mean();
+
+    referenceState.reinit(reference_state.size());
+    for(unsigned int i = 0 ; i < reference_state.size() ; i++){
+        referenceState(i) = reference_state(i);
+    }
+
+    snapshotMatrix = snapshotMatrix.colwise() - reference_state;
+
+    Eigen::BDCSVD<MatrixXd, Eigen::DecompositionOptions::ComputeThinU> svd(snapshotMatrix);
+    MatrixXd pod_basis = svd.matrixU();
+
+    fullBasis.reinit(pod_basis.rows(), pod_basis.cols());
+
+    for (unsigned int m = 0; m < pod_basis.rows(); m++) {
+        for (unsigned int n = 0; n < pod_basis.cols(); n++) {
+            fullBasis.set(m, n, pod_basis(m, n));
+        }
+    }
+
+    std::ofstream out_file("POD_basis.txt");
+    unsigned int precision = 16;
+    fullBasis.print_formatted(out_file, precision);
+
+    const Epetra_CrsMatrix epetra_system_matrix  = this->dg->system_matrix.trilinos_matrix();
+    Epetra_Map system_matrix_map = epetra_system_matrix.RowMap();
+    Epetra_CrsMatrix epetra_basis(Epetra_DataAccess::Copy, system_matrix_map, pod_basis.cols());
+
+    const int numMyElements = system_matrix_map.NumMyElements(); //Number of elements on the calling processor
+
+    for (int localRow = 0; localRow < numMyElements; ++localRow){
+        const int globalRow = system_matrix_map.GID(localRow);
+        for(int n = 0 ; n < pod_basis.cols() ; n++){
+            epetra_basis.InsertGlobalValues(globalRow, 1, &pod_basis(globalRow, n), &n);
+        }
+    }
+
+    Epetra_MpiComm epetra_comm(MPI_COMM_WORLD);
+    Epetra_Map domain_map((int)pod_basis.cols(), 0, epetra_comm);
+
+    epetra_basis.FillComplete(domain_map, system_matrix_map);
+
+    basis->reinit(epetra_basis);
+
+    return file_found;
+}
+
+template <int dim>
+std::shared_ptr<dealii::TrilinosWrappers::SparseMatrix> OfflinePOD<dim>::getPODBasis() {
```

Comment:
Were `getPODBasis()` and `getReferenceState()` written to make members `basis` and `referenceState` protected/private (they are currently public)?

Reply:
This is an example of polymorphism. I have two POD classes: OfflinePOD and OnlinePOD. Online POD takes snapshots on the fly and computes a POD basis for the adaptive sampling. Offline POD reads some previously computed snapshots stored on the hard drive and computes a POD basis. In both cases, the end use is to have a POD basis that I can use (and I also need a reference state to ensure consistency). Both POD classes implement the interface POD defined in pod_interface.h, which requires defining the functions getPODBasis() and getReferenceState(). The reduced-order ODE solvers only interface with POD classes using the functions defined in the interface POD. As a result, the reduced-order ODE solvers can use OfflinePOD or OnlinePOD, since both implement POD. Accessing data polymorphically like this can only be done with functions, not member variables.
It's also generally good practice to keep member variables private, like you said.